Replace overview when site is with a pending migration

renatho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #95379 and #95380

## Proposed Changes

* Extract some components from the "Hosting Features" tab, so we can reuse it.
* It customizes the Sites overview with a button to continue the migration when it's pending.
* The button contains a `ref` as `hosting-migration-overview`, so we can identify when the user is coming from this page.

#### This PR Doesn't Include

* Disable other tabs from the menu.
* Update the header when navigating to the site, similar to the changes we have in the listing.

@donnapep, do you think it's fine to deploy it as it is and implement these other parts in other PRs? I think one issue would be fine for both changes. I'll wait for your answer to then create the new issue, otherwise, I'll continue working on this one.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We want to offer a way for the users to continue their migration if they stopped in the middle.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* If https://github.com/Automattic/wp-calypso/pull/95413 wasn't merged yet, merge it into this branch to make tests more real and easier.
* Navigate to the flow `/setup/migration`.
* Navigate until the "How to migrate" step.
* Navigate to `/sites`.
* Click on the site, and click on "Start your migration".
* Make sure you navigate back to where you were in the flow.
* Click on the "Do it for me" option.
* Navigate again to the `/sites` and click on "Start your migration".
* Make sure you navigate back to your choice.
* Return one step and choose the "Do it myself".
* Navigate to the `/sites`.
* Check that you won't see the button because it changes quickly to the "started" status that will be implemented as part of https://github.com/Automattic/wp-calypso/issues/95383.
* Now repeat the tests using the flow `/setup/hosted-site-migration` and make sure it works properly.

## Screenshots

<img width="1399" alt="Screenshot 2024-10-18 at 18 40 00" src="https://github.com/user-attachments/assets/ec499f54-0781-46c9-b52d-19e7d5e5e741">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

renatho committed:
Fix bottom spacing in the content

Since the parent has an overflow hidden and a padding bottom it wasn't being respected.

renatho committed:
Extract hero components

renatho committed:
Move type to follow the same standard of the others

renatho committed:
Add key to the loop

renatho committed:
Extract grid component

renatho committed:
Add functions to get migration status and migration type

renatho commented on client/hosting/hosting-features/components/hosting-features.tsx:
```diff
@@ -28,7 +29,7 @@ type PromoCardProps = {
 };
 
 const PromoCard = ( { title, text, supportContext }: PromoCardProps ) => (
-	<HostingCard className="hosting-features__card" title={ title }>
```
Components were extracted so that we could reuse them.

renatho commented on client/hosting/hosting-features/components/hosting-features.tsx:
```diff
@@ -239,36 +238,35 @@ const HostingFeatures = () => {
 	return (
 		<div className="hosting-features">
-			<div className="hosting-features__hero">
+			<HostingHero>
```
Components were extracted so that we could reuse them.

renatho committed:
Tweak styles for when it does not have a link

renatho committed:
Add migration overview variation

renatho committed:
Temporarily add check for pending

When we implement the started status, we can also use the MigrationOverview, and remove this conditional.

renatho committed:
Redirect users to the proper step

renatho committed:
Redirect users to the proper step when the user didn't buy a plan yet

renatho committed:
Add query string to trigger the calypso_signup_start event

renatho commented:
> @donnapep, do you think it's fine to deploy it as it is and implement these other parts in other PRs? I think one issue would be fine for both changes. I'll wait for your answer to then create the new issue, otherwise, I'll continue working on this one.

I added the "[Status] String Freeze" because we have many new texts, so we'll delay a little to merge this anyway.

renatho commented on client/hosting/sites/components/dotcom-style.scss:
```diff
@@ -473,6 +473,7 @@
 			margin-top: 0;
 			margin-left: 0;
 			margin-right: 0;
+			margin-bottom: auto;
```
This fixes the spacing at the bottom of the tabs.

Without this, it wasn't respecting the padding.

<img width="1143" alt="Screenshot 2024-10-18 at 18 32 26" src="https://github.com/user-attachments/assets/7a788ed4-a5a1-4f4a-b2d6-28120e2debc6">

renatho committed:
Add tests for the links

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16927618

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @renatho for including a screenshot in the description! This is really helpful for our translators.

m1r0 commented on client/sites-dashboard/test/utils.js:
```diff
@@ -1,4 +1,5 @@
 import { isMigrationInProgress } from 'calypso/sites-dashboard/utils';
+import { getMigrationStatus, getMigrationType } from '../utils'; // Adjust the import path as necessary
```
The `isMigrationInProgress` import can be added here as well.

renatho replied:
Ah! Good catch! Fixed: https://github.com/Automattic/wp-calypso/pull/95497/commits/81f3cee6b410f4263012173f24091b06e961cf33

m1r0 reviewed (approved):
Looks good and works as described! Thanks for splitting those components.

I've added one small non-blocker suggestion.

donnapep commented:
@renatho Could you add some additional context or screen shots of what these mean? It's a bit unclear to me.

> Disable other tabs from the menu.
> Update the header when navigating to the site, similar to the changes we have in the listing.

renatho commented:
> @renatho Could you add some additional context or screen shots of what these mean? It's a bit unclear to me.

Sure! I was referring to these parts:

![Screenshot 2024-10-21 at 14 09 37](https://github.com/user-attachments/assets/fb62ea9a-953a-4964-bd5d-e9d833d35643)

renatho committed:
Join imports

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16927618).

donnapep commented:
@renatho Thank, I missed those subtleties. I can create an issue for adding the status to the header.

@fditrapani Can you confirm that we should disable the _Hosting settings_ tab for pending / started migrations?

fditrapani commented:
Thanks for the ping. This looks good from the screenshots. We had some back and forth about this with Michael. I figured it made sense to keep it disabled so you can stay focused on starting your migration. I can see the other side of it that people might want to poke around so I can be convinced otherwise. What do you all think?

Imran92 commented on client/hosting/overview/components/migration-overview.tsx:
```diff
@@ -0,0 +1,134 @@
+
+		let continueMigrationUrl;
+
+		if ( isFreePlan ) {
+			// For the flows where the checkout is after the choice.
+			switch ( migrationType ) {
+				case 'diy':
+					continueMigrationUrl = addQueryArgs(
+						{
+							...baseQueryArgs,
+							destination: 'upgrade',
+							how: 'myself',
+						},
+						'/setup/site-migration/site-migration-upgrade-plan'
+					);
+					break;
+				case 'difm':
+					continueMigrationUrl = addQueryArgs(
+						{
+							...baseQueryArgs,
+							destination: 'upgrade',
+							how: 'difm',
+						},
+						'/setup/site-migration/site-migration-upgrade-plan'
+					);
+					break;
+				default:
+					continueMigrationUrl = addQueryArgs(
+						baseQueryArgs,
+						'/setup/site-migration/site-migration-how-to-migrate'
+					);
+			}
```
Just a little concern, as we're only checking if the site is on a free plan or not, in case the site is on a `Personal` or `Premium` plan, will it send the user to the credentials or how-to-migrate step instead of the pre-checkout step?

renatho replied:
Hmm!! Good point, @Imran92! I was thinking that in our flows users wouldn't be able to pick a Personal or Premium plan and then try to migrate, but after you said, I tested it and I could do it through the `/start`, buying the plan (Personal or Premium), select the intent to migrate and use the flow.

I'll update it to a different check.

renatho replied:
Updated the check to match what we check in the flows to see if we'll redirect the user to the upgrade plan step or not. Now it makes more sense! Thank you for reviewing and catching this!

I also took the opportunity to refactor the duplicate code.

https://github.com/Automattic/wp-calypso/pull/95497/commits/3887d73cf58abd97025a2d745e6564fa5cc81d6d https://github.com/Automattic/wp-calypso/pull/95497/commits/2da505cadcab85fcba57cd756599270f2a7d7a18

m1r0 replied:
@renatho can you update the unit tests as well?

renatho replied:
Oops! Fixed here: https://github.com/Automattic/wp-calypso/pull/95497/commits/e5d856e258bcbb497c36a5665d59b4802c6261a9

renatho committed:
Extract canInstallPlugins logic

renatho committed:
Improve check to match what is checked in the flows

renatho committed:
Fix tests

renatho commented:
@m1r0 and @Imran92, since we had some iterations after the approval, I'll wait for a new approval, okay?

renatho committed:
Fix type to get from relative path

Imran92 commented on client/hosting/overview/components/migration-overview.tsx:
```diff
@@ -0,0 +1,132 @@
+				case 'diy':
+					continueMigrationUrl = addQueryArgs(
+						baseQueryArgs,
+						'/setup/migration/migration-how-to-migrate'
```
According to the [issue](https://github.com/Automattic/wp-calypso/issues/95380), in case of `migration-pending-diy` sticker, the user should be taken to the `site-migration-instructions` step. Looks like this here is taking us to the `migration-how-to-migrate` step. In case it's not intentional, should we fix it?

<img width="803" alt="image" src="https://github.com/user-attachments/assets/51b70579-fde4-43f3-b6d9-c213e2fda7d2">

renatho replied:
Good catch, @Imran92! Fixed in https://github.com/Automattic/wp-calypso/pull/95497/commits/2de7505061f8eca40813a2ca787f30d34491d2e3

renatho committed:
Fix DYI URL

Imran92 reviewed (commented):
I'm seeing a weird issue, but not sure why it's happening.

I had a site in a pending state. In the overview, section worked as expected, showing me correct button and everything relevant.

Then I changed the sticker to 'migration-completed-*'. But even after refreshing and hard-refreshing the page section many times, I still see the the migration-pending overview page.

https://github.com/user-attachments/assets/daa5ae91-ce05-474a-ae5a-cde82288df2c

But apart from that, LGTM. So I'm approving. Feel free to look into the issue either separately or here.

renatho commented:
@Imran92, there's a cache, but after 1 or 2 minutes it's updated. Could you confirm if it works for you after 1 or 2 minutes? I tested a few times and it's working well for me, just with this small cache.

renatho committed:
Remove not used conditional

renatho commented:
I just removed a not-needed conditional that I left accidentally: https://github.com/Automattic/wp-calypso/pull/95497/commits/456d3aa071b4c71b6057595fce550aca3bc1eb87

renatho committed:
Update tests to also mock the canInstallPlugins

Similar to the other mocks in these tests.

renatho commented:
Sorry, but I did 2 last minute changes that I noticed while working on another PR:

1. I removed a not-needed conditional that I left accidentally: https://github.com/Automattic/wp-calypso/commit/456d3aa071b4c71b6057595fce550aca3bc1eb87. It wasn't being used at all since we are checking this in the parent component.
2. I also changed the tests to also mock another dependency, similar to the others: https://github.com/Automattic/wp-calypso/pull/95497/commits/84a44ecc89d0520aeb420164b84bf9586b661975

Imran92 reviewed (approved):
LGTM 👍 Cache gets cleared after a couple of minutes

renatho commented:
I just rebased to fix a conflict, so I'm doing a sanity test, and I'll wait for the tests and merge it.

renatho committed:
Remove missed import after rebase
